cmds/dhclient/dhcp6_test.go

```diff
@@ -0,0 +1,362 @@
+	// Send IPv6 address with 60 second preferred lifetime,
+	// 90 second valid lifetime, no extra options
+	iaaddr, err := dhcp6.NewIAAddr(ip, 60*time.Second, 90*time.Second, nil)
+	if err != nil {
+		return err
+	}
+
+	// Add IAAddr inside IANA, add IANA to options
+	_ = ia.Options.Add(dhcp6.OptionIAAddr, iaaddr)
+	_ = w.Options().Add(dhcp6.OptionIANA, ia)
+
+	// Advertise address to soliciting clients
+	log.Printf("advertising IP: %s", ip)
+	_, err = w.Send(dhcp6.MessageTypeAdvertise)
+	return err
+}
+
+// First test
+// use tcpdump to get snapshots of packets to create []byte and
+// and then compare to packets you create and marshal.
+// TestDhcpClientRequest creates a packet using the dhcp6 package
+// and compares it to a pre-created packet to see if it is right.
+func TestDhcpClientRequest(t *testing.T) {
+	iface, err := ifup("enp0s25")
+	if err != nil {
+		t.Fatalf("failed to setup interface: %v\n", err)
+	}
+
+	mac := iface.Attrs().HardwareAddr
+	options, err := addOptions(mac)
+	if err != nil {
+		t.Fatalf("failed to add options: %v\n", err)
+	}
+
+	p := &dhcp6.Packet{
+		MessageType:   dhcp6.MessageTypeSolicit,
+		TransactionID: [3]byte{0x00, 0x01, 0x02},
+		Options:       options,
+	}
+
+	pb, err := p.MarshalBinary()
+	if err != nil {
+		t.Fatalf("packet marshal to binary err: %v\n", err)
+	}
+
+	h1 := &ipv6.Header{
+		Version:      ipv6.Version,
+		TrafficClass: 0,
+		FlowLabel:    rand.Int() & 0xfffff,
+		PayloadLen:   udpHdrLen + len(pb),
+		NextHeader:   unix.IPPROTO_UDP,
+		HopLimit:     1,
+		// TODO: src ip harded coded for now
+		Src: net.ParseIP("fe80::baae:edff:fe79:6191"), // net.ParseIP("fe80::179a:1422:c923:2727"),
+		Dst: net.ParseIP("FF02::1:2"),
+	}
+
+	h2 := &Udphdr{
+		Src:    srcPort,
+		Dst:    dstPort,
+		Length: uint16(udpHdrLen + len(pb)),
+	}
+
+	pkt, err := marshalPacket(h1, h2, pb)
+	if err != nil {
+		t.Fatalf("failed to make a new packet: %v\n", err)
+	}
+
+	// Test if identical after marshal and unmarshal
+	h1p, h2p, _, err := unmarshalPacket(pkt)
+	if err != nil {
+		t.Fatalf("failed to unmarshal packet: %v\n", err)
+	}
+	if !reflect.DeepEqual(h1, h1p) {
+		t.Fatalf("ip headers do not match:\n%v\n%v\n", h1, h1p)
+	}
+	if !reflect.DeepEqual(h2, h2p) {
+		t.Fatalf("udp headers do not match:\n%v\n%v\n", h2, h2p)
+	}
+	// if !reflect.DeepEqual(p, pp) {
+	// 	t.Fatalf("dhcp packets do not match:\n%v\n%v\n", p, pp)
+	// }
+
+	// Test if identical with valid packet
+	// validPkt = []byte{}
+
+}
+
+// Second test. Try to talk to the server and see what comes
+// back as errors.
+// TestDhcp6ClientErrors creates requests with errors and checks
+// for what we think are correct error responses from the server.
+func TestDhcp6ClientErrors(t *testing.T) {
+	if os.Getuid() != 0 {
+		t.Skip("we need root to serve dhcp6")
+	}
+
+	if err := serverHelper(t); err != nil {
+		t.Fatalf("starting server helper; want nil, got %v", err)
+	}
+
+	iface, err := ifup("enp0s25")
```

Comment:
This is going to be a problem, we'll need to fix later, because enp0s25 is a name that won't be the same on many hosts.